lib/msal-core/src/ScopeSet.ts

```diff
@@ -160,4 +153,93 @@ export class ScopeSet {
+        let oidcScopesFound = 0;
+
+        if (scopes.indexOf(Constants.openidScope) > -1) {
+            oidcScopesFound += 1;
+        }
+
+        if (scopes.indexOf(Constants.profileScope) > -1) {
+            oidcScopesFound += 1;
+        }
+
+        const nonOidcScopes = scopes.length - oidcScopesFound;
+
+        return (nonOidcScopes === 0);
+
+    }
+
+       /**
+     * @ignore
+     * Returns true if the scopes array only contains openid and/or profile
+     */
+    static containsAnyOidcScopes(scopes: Array<string>): boolean {
+        const containsOpenIdScope = scopes.indexOf(Constants.openidScope) > -1;
+        const containsProfileScope = scopes.indexOf(Constants.profileScope) > -1;
+
+        return (containsOpenIdScope || containsProfileScope);
+    }
+
+    /**
+     * @ignore
+     * Returns true if the clientId is the only scope in the array
+     */
+    static onlyContainsClientId(scopes: Array<String>, clientId: string): boolean {
+        const containsClientId = scopes.indexOf(clientId) > -1;
+        const containsSingleScope = scopes.length === 1;
+
+        return containsClientId && containsSingleScope;
+    }
+
+    /**
+     * @ignore
+     * Normalizes scopes array before creating URL string by removing clientId if present and
+     * appending openid and/or profile whenever they are missing.
+     * @param scopes Array<string>: Pre-normalized scopes array
+     * @param clientId string: The application's clientId that is searched for in the scopes array
+     */
+    static normalize(scopes: Array<string>, clientId: string): Array<string> {
+        const prunedScopes = this.spliceClientIdIfSingleScope(scopes, clientId);
+        return this.appendDefaultScopes(prunedScopes);
+    }
+
+    /**
+     * @ignore
+     * Adds missing OIDC scopes to scopes array withot duplication. Since STS requires OIDC scopes for
+     * all implicit flow requests, 'openid' and 'profile' should always be included in the final request
+     */
+    private static appendDefaultScopes(scopes: Array<string>): Array<string> {
+        const extendedScopes = scopes;
+        if (extendedScopes.indexOf(Constants.openidScope) === -1) {
+            extendedScopes.push(Constants.openidScope);
+        }
+
+        if(extendedScopes.indexOf(Constants.profileScope) === -1) {
+            extendedScopes.push(Constants.profileScope);
+        }
+
+        return extendedScopes;
+    }
+
+    /**
+     * @ignore
+     * Removes clientId from scopes array if included as only scope. If it's not the only scope, it is treated as a resource scope.
+     * @param scopes Array<string>: Pre-normalized scopes array
+     * @param clientId string: The application's clientId that is searched for in the scopes array
+     */
+    private static spliceClientIdIfSingleScope(scopes: Array<string>, clientId: string): Array<string> {
+        const clientIdIndex: number = scopes.indexOf(clientId);
```

Comment:
you can simplify this function by doing something like 

`return this.onlyContainsClientId(scopes, clientId)? [] : scopes`

Reply:
Awesome! I had a feeling there was a way to clean this up that I hadn't though of. This is exactly what I was looking for!

Reply:
Ok, ended up implementing this in the renamed translateClientIdIfSingleScope, only instead of an empty array we now actually return the OIDC scopes array if clientId is the only scope.